mariamihaelafainarea/assignment3/Assignment3

```diff
@@ -0,0 +1,171 @@
+                }
+            }
+        }
+        
+    }
+    //A dictionary based on a trie ! 
+    //It has a better complexity than
+    //a hashmap !
+    class MyDictionary{
+        private class TrieNode{
+            private HashMap<Character,TrieNode> children;
+            private boolean endOfWord;
+            public TrieNode() {
+                children = new HashMap<>();
+                endOfWord = false;
+            }
+            
+        }
+        private final TrieNode root;
+        //the idea is very simple
+        //you start from root
+        //you go deep down in the trie 
+        //with the current node
+        //if theChildren (the character that need to be added)
+        //exists you only pass it and go deep down
+        public void insert(String word) {
+            TrieNode current = root;
+            for(int i = 0; i < word.length();i++) {
+                TrieNode theChild = current.children.get(word.charAt(i));
+                if (theChild == null) {  
+                    theChild = new TrieNode();
+                    current.children.put(word.charAt(i), theChild);
+                }
+                current = theChild;
+            }
+            //after you add the last node and move your
+            //current there
+            //you mark it as endOfWord
+            current.endOfWord = true;
+            
+        }
+        public boolean isGeneralHelper(String action,TrieNode current,String word,int index) {
+            if(action.equals("word")) {
+                if (index == word.length()) {
+                    return current.endOfWord;
+                }
+            } else {
+                if (index == word.length()) {
+                    if(current == null) {
+                        return false;
+                    }
+                return true;
+                }
+            }
+            
+            TrieNode theChildren = current.children.get(word.charAt(index));
+            if (theChildren == null) {
+               return false;
+            }
+            index ++;
+            return isGeneralHelper(action,theChildren, word, index);
+            
+        }
+
+        //it will call a recursive function that will 
+        //try at each step to find in the children
+        //of the node expanded the character from the word
+        //the auxiliary function get into recursion
+        //with a cut word (shorter) and wait 
+        //the end of the word
+        //to return the value of endOfWord
+        public boolean isWord(String word) {
+           return isGeneralHelper("word",root,word,0);
+        }
+  
+        //the idea is similar with the isWord but here
+        //you need a node in return.
+        //it helps me to know if there
+        //something exists and is not the end/ is the end = > prefix
+        //something doesn't exists => not a prefix
+        public boolean isPrefix(String s) {
+            boolean x = isGeneralHelper("prefix",root,s,0);
```

Comment:
You don't need the x variable at all. Just immediately return the value of the method call.

Reply:
done